File: Aufgabe1/solutions/1.2/sqrt.cpp

#include <iostream>
#include <cmath>

double sqrt_newton(double x) {
  const double       EPSILON  = 0.000001;
  const unsigned int MAX_ITER = 10000;

  unsigned int iter     = 0;
  double       estimate = 1.0;

  while (::fabs(estimate*estimate - x) > EPSILON && iter < MAX_ITER) {
    estimate = 0.5*(estimate + x/estimate);
    ++iter;
  }

  return estimate;
}

int main() {
  double x = 0.0;
  std::cout << "Bitte geben Sie eine Zahl ein: ";
  std::cin >> x;
 
  std::cout << "Die Wurzel aus " << x << " ist " << sqrt_newton(x) << std::endl;
 
  return 0;
}

File: Aufgabe1/solutions/1.2/sqrt.hs
import System.IO

-- Fundamental expressions for solving the problem
isGoodEnough x y = abs (y*y - x) < 0.0001
improve x y = 0.5*(y + x/y)

-- Solution 1
sqrt_newton x = head (filter (isGoodEnough x) (iterate (improve x) 1))

-- Solution 2
-- Note: this does not use the global iterate from the 
-- standard library!
sqrt_newton' x = iterate x 1
  where iterate x y | (isGoodEnough x y) = y
                    | otherwise = iterate x (improve x y)

-- Program entry point
main :: IO ()
main = putStr "Bitte geben Sie eine Zahl ein: " >>
  hFlush stdout >>
  getLine >>= 
  (\xStr -> 
    let x = read xStr :: Double
    in putStr ("Die Wurzel aus " ++ xStr ++ " ist: ") >> putStrLn (show $ sqrt_newton' x))